File: exercises-cpp/james_clarke/ex02_oo_basics/src/ShapeSorter.cpp
#include "ShapeSorter.h"

ShapeSorter::ShapeSorter(vector<shared_ptr<Shape>> shapes) {
	_shapes = shapes;
}

void ShapeSorter::OfType(string type) {
	for (const auto& s:_shapes)
	{
		if (s->Type() == type)
			cout << s->Str() << endl;
	}
}

void ShapeSorter::WithSides(int numberOfSides) {
	int matchingShapes = 0;
	for (const auto& s:_shapes)
	{
		if (s->Sides() == numberOfSides) {
			cout << s->Str() << endl;
			matchingShapes++;
		}
	}
	if (matchingShapes == 0)
		cout << printf("No shapes with %i sides", numberOfSides) << endl;
}

void ShapeSorter::AreaDescending() {
	auto sortedVector = copyPointers();
	sort(sortedVector.begin(), sortedVector.end(), [](shared_ptr<Shape> a, shared_ptr<Shape> b) { return a->Area() > b->Area(); });
	for (const auto& s : sortedVector)
		cout << s->Str() << endl;
}

void ShapeSorter::PerimeterDescending() {
	auto sortedVector = copyPointers();
	sort(sortedVector.begin(), sortedVector.end(), [](shared_ptr<Shape> a, shared_ptr<Shape> b) { return a->Perimeter() > b->Perimeter(); });
	for (const auto& s : sortedVector)
		cout << s->Str() << endl;
}

vector<shared_ptr<Shape>> ShapeSorter::copyPointers() {
	vector<shared_ptr<Shape>> newVector;
	for (const auto& s : _shapes)
		newVector.push_back(shared_ptr<Shape>(s));
	return newVector;
}

File: exercises-python/james_clarke/ex02_oo_basics/ShapeSorter.py
from Shapes import *


class ShapeSorter:
    def __init__(self, shapes: list[Shape]):
        self.shapes = shapes

    def of_type(self, shape_type: type):
        matching_shapes = [s for s in self.shapes if isinstance(s, shape_type)]
        if len(matching_shapes) == 0:
            print("No shapes of type " + str(shape_type))
            return
        for shape in matching_shapes:
            print(shape.to_string())

    def with_sides(self, number_sides: int):
        matching_shapes = [s for s in self.shapes if s.sides == number_sides]
        if len(matching_shapes) == 0:
            print("No shapes with " + str(number_sides) + " sides")
            return
        for shape in matching_shapes:
            print(shape.to_string())

    def area_descending(self):
        sorted_by_area = sorted(self.shapes, key=lambda s: s.area, reverse=True)
        for shape in sorted_by_area:
            print(shape.to_string())

    def perimeter_descending(self):
        sorted_by_perimeter = sorted(self.shapes, key=lambda s: s.perimeter, reverse=True)
        for shape in sorted_by_perimeter:
            print(shape.to_string())
